File: app/src/main/java/com/pr0gramm/app/services/Update.java
package com.pr0gramm.app.services;

import android.os.Parcel;
import android.os.Parcelable;

import org.immutables.gson.Gson;
import org.immutables.value.Value;

/**
 * Update
 */
@Value.Immutable
@Gson.TypeAdapters
public abstract class Update implements Parcelable {
    public abstract int version();

    public abstract String apk();

    public abstract String changelog();


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(version());
        dest.writeString(apk());
        dest.writeString(changelog());
    }

    public static final Creator<Update> CREATOR = new Creator<Update>() {
        public Update createFromParcel(Parcel source) {
            return ImmutableUpdate.builder()
                    .version(source.readInt())
                    .apk(source.readString())
                    .changelog(source.readString())
                    .build();
        }

        public Update[] newArray(int size) {
            return new Update[size];
        }
    };
}

File: app/src/main/java/com/pr0gramm/app/services/Update.kt
package com.pr0gramm.app.services

import android.os.Parcel
import com.pr0gramm.app.model.update.UpdateModel
import com.pr0gramm.app.parcel.DefaultParcelable
import com.pr0gramm.app.parcel.SimpleCreator
import com.pr0gramm.app.parcel.readStringNotNull

/**
 * Update
 */
data class Update(val version: Int, val apk: String, val changelog: String) : DefaultParcelable {
    constructor(update: UpdateModel) : this(update.version, update.apk, update.changelog)

    val versionStr = "1.${version / 10}.${version % 10}"

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(version)
        dest.writeString(apk)
        dest.writeString(changelog)
    }

    companion object CREATOR : SimpleCreator<Update>() {
        override fun createFromParcel(source: Parcel): Update {
            return Update(
                    version = source.readInt(),
                    apk = source.readStringNotNull(),
                    changelog = source.readStringNotNull(),
            )
        }
    }
}
